sailor/sap_iot/fetch_aggregates.py

```diff
@@ -0,0 +1,223 @@
+    # That is confusing, as it's neither a purely horizonal format nor a purely vertical format.
+    # Much of the code below is therefore taking care of conversion from the SAP IoT format to our own format
+    # which is fully horizontal (ie column headers encode template, indicator group, indicator and aggregation function)
+    # SAP IoT format is effectively:
+    # time | equi - id | equi - model - id | indicator - group - id | template - id | indicator_A_value_AGG
+    # x | equi - 1 | model - 1 | indi - group - 1 | template - 1 | 5
+    # x | equi - 1 | model - 1 | indi - group - 2 | template - 1 | 15
+    #
+    # Our format is effectively:
+    # time | equi - id | equi - model - id | column_id_1 | columnd_id_2
+    # x | equi - 1 | model - 1 | 5 | 15
+    # where the column_ids encode all unique identifiers of an aggregated sensor value here
+    # column_id = hash(indicator_id + indicator_group_id + template_id + aggregation_function)
+    if start is None or end is None:
+        raise ValueError("Time parameters must be specified")
+
+    start = _any_to_timestamp(start)
+    end = _any_to_timestamp(end)
+
+    query_groups = defaultdict(list)
+    for indicator in indicator_set:
+        query_groups[(indicator.template_id, indicator._liot_group_id)].append(indicator._liot_id)
+
+    if aggregation_functions is None:
+        aggregation_functions = _ALL_KNOWN_AGGREGATION_FUNCTIONS
+    aggregated_indicators = AggregatedIndicatorSet._from_indicator_set_and_aggregation_functions(indicator_set,
+                                                                                                 aggregation_functions)
+    oauth_iot = get_oauth_client('sap_iot')
+    df = pd.DataFrame({'equipment_id': [], 'timestamp': pd.to_datetime([], utc=True)})
+    duration = None
+
+    aggregation_interval = _parse_aggregation_interval(aggregation_interval)
+    params = dict(start=_timestamp_to_isoformat(start, True), end=_timestamp_to_isoformat(end, True),
+                  equipment_set=equipment_set, aggregated_indicator_set=aggregated_indicators,
+                  aggregation_interval=aggregation_interval)
+
+    # doing this as the aggregates api needs to be called for each indicator group
+    for (template_id, liot_group_id), group_indicators in query_groups.items():
+        LOG.debug("Fetching indicator data. Indicator Group ID: %s.", liot_group_id)
+        results = _fetch_aggregates(liot_group_id, template_id, oauth_iot, **params)
+        results_df = _prepare_df(results, aggregated_indicators, liot_group_id, template_id)
+        if results_df.empty:
+            continue
+        duration = results[0]['properties']['duration'] if duration is None else duration
+        df = pd.merge(df, results_df, on=['timestamp', 'equipment_id'], how='outer')
+
+    if df is None:
+        raise RuntimeError("No data could be found for the given parameters.")
```

Comment:
This check actually does not work. DF is always set. And when the for loop doesn't return any data, `duration` is None and it runs into an error in line 151 (TypeError: can only concatenate str (not "NoneType") to str).

This also warrants a test for the case that "no data is found".

And we should handle it differently. In other areas when no data is found, we don't throw an error. Instead we throw a warning and return an empty result.